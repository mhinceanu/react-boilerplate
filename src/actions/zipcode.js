import { SEARCH_ZIPCODE } from '../constants/actionTypes';
import api from '../api';

export const searchZipcode = (zipcode) => (dispatch) => {
  return dispatch(
    api.fetch(SEARCH_ZIPCODE, api.config.searchZipcode(zipcode), api.http.get())
  );
};
