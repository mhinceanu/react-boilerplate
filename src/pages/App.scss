@import '../styles/colors.scss';
@import '../styles/typography.scss';

.bp-app {
    align-items: center;
    background-color: $violet;
    color: white;
    display: flex;
    flex-direction: column;
    height: 100vh;
    padding: 0 20px;

    > h4 {
        font-size: $h4-mobile;
        margin-top: 100px;
    }

    > p {
        font-size: $p-mobile;
        margin-top: 10px;
    }

    &__search {
        align-items: center;
        display: flex;
        flex-direction: column;
        width: clamp(300px, 100%, 500px);

        > input {
            background-color: rgba(0, 0, 0, 0);
            border: none;
            border-bottom: 1px solid $caramel;
            color: $tea;
            font-size: 1.6rem;
            font-weight: 600;
            height: 40px;
            margin: 20px 0;
            text-align: center;
            width: 100%;
        }

        > button {
            height: 40px;
            width: 200px;
        }
    }
}