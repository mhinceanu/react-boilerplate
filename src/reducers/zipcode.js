import { SEARCH_ZIPCODE } from '../constants/actionTypes';
import { PENDING, SUCCESS, ERROR } from '../constants/actionStatus';

export const initialState = {
  isFetching: false,
  fetched: false,
  data: null,
  error: null
};

const zipcode = (state = initialState, action = {}) => {
  const { type, response, error } = action;

  switch (type) {
    case `${SEARCH_ZIPCODE}_${PENDING}`:
      return {
        ...state,
        isFetching: true,
        fetched: false,
        data: null,
        error: null
      };

    case `${SEARCH_ZIPCODE}_${SUCCESS}`:
      return {
        ...state,
        isFetching: false,
        fetched: true,
        data: response
      };

    case `${SEARCH_ZIPCODE}_${ERROR}`:
      return {
        ...state,
        isFetching: false,
        fetched: true,
        error
      };

    default:
      return state;
  }
};

export default zipcode;
