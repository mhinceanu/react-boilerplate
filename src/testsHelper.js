import React from 'react';
import createStore from './store/createStore';

export const WrapStore = ({ children, initialState = {} }) => {
  const state = { ...initialState };
  const storeConfig = createStore(state);

  return (
    <storeConfig.Provider store={storeConfig.store}>
      {children}
    </storeConfig.Provider>
  );
};
