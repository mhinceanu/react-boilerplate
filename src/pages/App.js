import React, { useCallback, useMemo, useState } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import LoadingIndicator from '../shared/LoadingIndicator';
import ZipcodeDetails from '../components/ZipcodeDetails';
import { searchZipcode } from '../actions/zipcode';
import './App.scss';

function App() {
  const dispatch = useDispatch();
  const [value, setValue] = useState('');

  const zipcodeState = useSelector((state) => state.zipcode, shallowEqual);

  const handleChange = useCallback((e) => {
    const inputValue = e.target.value;
    setValue(inputValue);
  }, []);

  const handleSearch = useCallback(() => {
    dispatch(searchZipcode(Number(value)));
  }, [dispatch, value]);

  const handleKeyUp = useCallback(
    (e) => {
      if (e.key === 'Enter') {
        handleSearch();
      }
    },
    [handleSearch]
  );

  const isDisabled = useMemo(() => {
    return value?.length === 0;
  }, [value]);

  return (
    <div className="bp-app">
      <h4>Search information about zipcode.</h4>
      <p>Here are some examples: 90210, 33162</p>

      <div className="bp-app__search">
        <input
          tabIndex={0}
          type="number"
          placeholder="Zipcode..."
          value={value}
          onChange={(e) => handleChange(e)}
          onKeyUp={handleKeyUp}
        />
        <button
          tabIndex={0}
          onClick={handleSearch}
          onKeyUp={handleSearch}
          disabled={isDisabled}
        >
          Search
        </button>
      </div>

      {zipcodeState.isFetching && <LoadingIndicator />}

      {zipcodeState.fetched && <ZipcodeDetails details={zipcodeState.data} />}
    </div>
  );
}

export default App;
