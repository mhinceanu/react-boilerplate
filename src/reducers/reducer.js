import zipcode from './zipcode';

export const reducers = {
  zipcode
};
