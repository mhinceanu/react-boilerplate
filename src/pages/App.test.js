import React from 'react';
import { render } from '@testing-library/react';
import { WrapStore } from '../testsHelper';
import { initialState } from '../reducers/zipcode';
import App from './App';

describe('<App />', function () {
  const getComponent = () => {
    return (
      <WrapStore initialState={{ zipcode: initialState }}>
        <App />
      </WrapStore>
    );
  };

  it('should render <App />', () => {
    const { container } = render(getComponent());

    expect(container).toMatchSnapshot();
  });
});
