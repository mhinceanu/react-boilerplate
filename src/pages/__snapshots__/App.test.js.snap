// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render <App /> 1`] = `
<div>
  <div
    class="bp-app"
  >
    <h4>
      Search information about zipcode.
    </h4>
    <p>
      Here are some examples: 90210, 33162
    </p>
    <div
      class="bp-app__search"
    >
      <input
        placeholder="Zipcode..."
        tabindex="0"
        type="number"
        value=""
      />
      <button
        disabled=""
        tabindex="0"
      >
        Search
      </button>
    </div>
  </div>
</div>
`;
